### DISCUSSION TITLE
why is IPC less than 1?

### DISCUSSION THREAD

**Discussion by icuuuuuuer** [ChampSim/ChampSim#542] [Q&A] [answered]
I am currently conducting research on the impact of DRAM bandwidth on the overall CPU system. During my investigation, I encountered an interesting phenomenon, which I have summarized into three questions:
1.When I set the CPU parameters quite aggressively and run 44 memory-intensive traces from SPEC2017, the geometric mean IPC still does not exceed 1. This seems counterintuitive, as an 8-wide issue CPU should typically have an IPC greater than 1. Could there be something unreasonable in my parameter settings? Below are my configurations, and I would greatly appreciate it if any experienced Champsim users could take a look and provide some feedback.
2.Is the following method correct for calculating the theoretical DRAM bandwidth? The amount of data (in GB) transferred per cycle: 8 / 1024 * 1024 * 1024, and the time per cycle (in seconds): 1 / 2554000000. When dividing GB by the time in seconds, the calculated DDR bandwidth comes out to be 19GB/s. Is this calculation accurate?
3.Lastly, can increasing or decreasing DDR bandwidth be achieved simply by adjusting the frequency parameter?
Thank you very much in advance for your time and assistance!

Traces:
```
602.gcc_s-1850B.champsimtrace.xz
602.gcc_s-2226B.champsimtrace.xz
602.gcc_s-2375B.champsimtrace.xz
602.gcc_s-734B.champsimtrace.xz
603.bwaves_s-1740B.champsimtrace.xz
603.bwaves_s-2609B.champsimtrace.xz
603.bwaves_s-2931B.champsimtrace.xz
603.bwaves_s-891B.champsimtrace.xz
605.mcf_s-1152B.champsimtrace.xz
605.mcf_s-1536B.champsimtrace.xz
605.mcf_s-1554B.champsimtrace.xz
605.mcf_s-1644B.champsimtrace.xz
605.mcf_s-472B.champsimtrace.xz
605.mcf_s-484B.champsimtrace.xz
605.mcf_s-665B.champsimtrace.xz
605.mcf_s-782B.champsimtrace.xz
605.mcf_s-994B.champsimtrace.xz
607.cactuBSSN_s-2421B.champsimtrace.xz
607.cactuBSSN_s-3477B.champsimtrace.xz
607.cactuBSSN_s-4004B.champsimtrace.xz
619.lbm_s-2676B.champsimtrace.xz
619.lbm_s-2677B.champsimtrace.xz
619.lbm_s-3766B.champsimtrace.xz
619.lbm_s-4268B.champsimtrace.xz
620.omnetpp_s-141B.champsimtrace.xz
620.omnetpp_s-874B.champsimtrace.xz
621.wrf_s-6673B.champsimtrace.xz
621.wrf_s-8065B.champsimtrace.xz
623.xalancbmk_s-10B.champsimtrace.xz
623.xalancbmk_s-165B.champsimtrace.xz
623.xalancbmk_s-202B.champsimtrace.xz
627.cam4_s-490B.champsimtrace.xz
628.pop2_s-17B.champsimtrace.xz
649.fotonik3d_s-10881B.champsimtrace.xz
649.fotonik3d_s-1176B.champsimtrace.xz
649.fotonik3d_s-7084B.champsimtrace.xz
649.fotonik3d_s-8225B.champsimtrace.xz
654.roms_s-1007B.champsimtrace.xz
654.roms_s-1070B.champsimtrace.xz
654.roms_s-1390B.champsimtrace.xz
654.roms_s-293B.champsimtrace.xz
654.roms_s-294B.champsimtrace.xz
654.roms_s-523B.champsimtrace.xz
657.xz_s-2302B.champsimtrace.xz
```
Results:
``` results
602.gcc_s-1850B.champsimtrace_output.txt: 2.59 IPC, Accuracy: None, Timeliness: None, Bandwidth: 11.0 GB/S
602.gcc_s-2226B.champsimtrace_output.txt: 0.8119 IPC, Accuracy: None, Timeliness: None, Bandwidth: 13.47 GB/S
602.gcc_s-2375B.champsimtrace_output.txt: 1.593 IPC, Accuracy: None, Timeliness: None, Bandwidth: 0.47 GB/S
602.gcc_s-734B.champsimtrace_output.txt: 2.679 IPC, Accuracy: None, Timeliness: None, Bandwidth: 5.99 GB/S
603.bwaves_s-1740B.champsimtrace_output.txt: 0.7039 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.28 GB/S
603.bwaves_s-2609B.champsimtrace_output.txt: 0.7017 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.26 GB/S
603.bwaves_s-2931B.champsimtrace_output.txt: 0.6283 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.05 GB/S
603.bwaves_s-891B.champsimtrace_output.txt: 3.73 IPC, Accuracy: None, Timeliness: None, Bandwidth: 13.87 GB/S
605.mcf_s-1152B.champsimtrace_output.txt: 0.3592 IPC, Accuracy: None, Timeliness: None, Bandwidth: 1.6 GB/S
605.mcf_s-1536B.champsimtrace_output.txt: 0.4634 IPC, Accuracy: None, Timeliness: None, Bandwidth: 5.43 GB/S
605.mcf_s-1554B.champsimtrace_output.txt: 0.6155 IPC, Accuracy: None, Timeliness: None, Bandwidth: 5.13 GB/S
605.mcf_s-1644B.champsimtrace_output.txt: 0.1253 IPC, Accuracy: None, Timeliness: None, Bandwidth: 4.42 GB/S
605.mcf_s-472B.champsimtrace_output.txt: 0.6634 IPC, Accuracy: None, Timeliness: None, Bandwidth: 5.89 GB/S
605.mcf_s-484B.champsimtrace_output.txt: 0.6601 IPC, Accuracy: None, Timeliness: None, Bandwidth: 5.25 GB/S
605.mcf_s-665B.champsimtrace_output.txt: 0.4465 IPC, Accuracy: None, Timeliness: None, Bandwidth: 1.3 GB/S
605.mcf_s-782B.champsimtrace_output.txt: 0.1199 IPC, Accuracy: None, Timeliness: None, Bandwidth: 2.04 GB/S
605.mcf_s-994B.champsimtrace_output.txt: 0.5694 IPC, Accuracy: None, Timeliness: None, Bandwidth: 2.13 GB/S
607.cactuBSSN_s-2421B.champsimtrace_output.txt: 1.463 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.14 GB/S
607.cactuBSSN_s-3477B.champsimtrace_output.txt: 3.051 IPC, Accuracy: None, Timeliness: None, Bandwidth: 2.39 GB/S
607.cactuBSSN_s-4004B.champsimtrace_output.txt: 2.802 IPC, Accuracy: None, Timeliness: None, Bandwidth: 2.3 GB/S
619.lbm_s-2676B.champsimtrace_output.txt: 0.2944 IPC, Accuracy: None, Timeliness: None, Bandwidth: 4.08 GB/S
619.lbm_s-2677B.champsimtrace_output.txt: 0.2082 IPC, Accuracy: None, Timeliness: None, Bandwidth: 4.15 GB/S
619.lbm_s-3766B.champsimtrace_output.txt: 0.4426 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.97 GB/S
619.lbm_s-4268B.champsimtrace_output.txt: 0.2086 IPC, Accuracy: None, Timeliness: None, Bandwidth: 4.14 GB/S
620.omnetpp_s-141B.champsimtrace_output.txt: 0.8806 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.06 GB/S
620.omnetpp_s-874B.champsimtrace_output.txt: 0.9511 IPC, Accuracy: None, Timeliness: None, Bandwidth: 2.98 GB/S
621.wrf_s-6673B.champsimtrace_output.txt: 0.5347 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.78 GB/S
621.wrf_s-8065B.champsimtrace_output.txt: 1.684 IPC, Accuracy: None, Timeliness: None, Bandwidth: 4.09 GB/S
623.xalancbmk_s-10B.champsimtrace_output.txt: 1.86 IPC, Accuracy: None, Timeliness: None, Bandwidth: 4.16 GB/S
623.xalancbmk_s-165B.champsimtrace_output.txt: 4.668 IPC, Accuracy: None, Timeliness: None, Bandwidth: 1.75 GB/S
623.xalancbmk_s-202B.champsimtrace_output.txt: 3.103 IPC, Accuracy: None, Timeliness: None, Bandwidth: 10.54 GB/S
627.cam4_s-490B.champsimtrace_output.txt: 2.472 IPC, Accuracy: None, Timeliness: None, Bandwidth: 2.71 GB/S
628.pop2_s-17B.champsimtrace_output.txt: 3.002 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.55 GB/S
649.fotonik3d_s-10881B.champsimtrace_output.txt: 0.6033 IPC, Accuracy: None, Timeliness: None, Bandwidth: 4.11 GB/S
649.fotonik3d_s-1176B.champsimtrace_output.txt: 1.323 IPC, Accuracy: None, Timeliness: None, Bandwidth: 5.54 GB/S
649.fotonik3d_s-7084B.champsimtrace_output.txt: 0.915 IPC, Accuracy: None, Timeliness: None, Bandwidth: 4.71 GB/S
649.fotonik3d_s-8225B.champsimtrace_output.txt: 1.314 IPC, Accuracy: None, Timeliness: None, Bandwidth: 5.5 GB/S
654.roms_s-1007B.champsimtrace_output.txt: 1.334 IPC, Accuracy: None, Timeliness: None, Bandwidth: 1.69 GB/S
654.roms_s-1070B.champsimtrace_output.txt: 0.6004 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.89 GB/S
654.roms_s-1390B.champsimtrace_output.txt: 0.8392 IPC, Accuracy: None, Timeliness: None, Bandwidth: 4.66 GB/S
654.roms_s-293B.champsimtrace_output.txt: 2.496 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.93 GB/S
654.roms_s-294B.champsimtrace_output.txt: 1.325 IPC, Accuracy: None, Timeliness: None, Bandwidth: 3.82 GB/S
654.roms_s-523B.champsimtrace_output.txt: 0.9354 IPC, Accuracy: None, Timeliness: None, Bandwidth: 5.51 GB/S
657.xz_s-2302B.champsimtrace_output.txt: 1.203 IPC, Accuracy: None, Timeliness: None, Bandwidth: 0.8 GB/S
geomean:0.9322
```
```
bin/champsim --warmup_instructions 50000000 --simulation_instructions 50000000  ./traces
```

**Comment by maccoymerrell** [comment]
1. What exactly are you altering? For memory intensive traces, the memory system itself will be the bottleneck rather than the core. Cache and dram parameters should be your focus. An IPC of less than 1 for memory intensive traces isn't uncommon. It can take hundreds of cycles to serve a cache miss, which can translate to very low IPCs irrespective of issue or retire width.
2. That is roughly correct. The peak bandwidth that the data bus is capable of is frequency * channel width.
3. To increase memory bandwidth, there are a few ways. 
A. Increase the # of memory channels. These operate in parallel, so bandwidth doubles with 2 channels, and quadruples with 4 channels.
B. Increase the data bus frequency. More clocks per second == more data per second
C. Increase the channel width. More data per clock. This is something that is pretty standardized for ddr, so I can't recommend messing with this too much.

**Reply by icuuuuuuer** [reply]
I initially used the default CPU core configuration in Champsim. However, when I adjusted the frequency to increase DDR bandwidth from 5GB/s to 7GB/s, 9GB/s, 11GB/s, and 13GB/s, I noticed that IPC growth was minimal. This suggests that the core might be somewhat limited, with insufficient memory access parallelism. My goal, however, is to make DDR bandwidth as close as possible to being the system’s main bottleneck, which is why I adopted this more aggressive 8-issue CPU configuration. Do you have any additional suggestions?
Regarding the second point, you mentioned it was 'roughly correct.' Is there a more accurate way to calculate DDR bandwidth? It would help me achieve more precise results. Thank you very much !

**Reply by maccoymerrell** [reply, marked as answer]
A couple of things.
1. To make dram bandwidth the main bottleneck, the workloads will need to be non-cache residential. Some workloads are already like this, others are not. Reducing cache size would exacerbate this effect.
2. The caches themselves should not be more dominant towards latency than the dram itself. Make sure tag checking and fill bandwidth is high at all levels of the cache.
3. Core parameters are going to have minimal effect. If you are memory constrained, then many instructions are probably waiting in the core on data that is in flight. Wider widths, more functional units, etc... is not going to help much. At the same time, dram bandwidth may not be utilized. Think about what happens if a workload can reside entirely in the LLC. You will still get low IPC (probably less than 1) but have low dram bandwidth utilization. Caching is probably going to be the main problem with relating dram bandwidth with cpu performance.

When I say roughly correct, I mean that in an ideal world where no row activations are needed, dram refresh isn't a thing, etc... You will find that often times banks may be occupied with maintenance requests like refresh. Since those are issued at time intervals, they technically reduce your maximum bandwidth by the time they reserve in the banks and the time it takes the memory controller to issue them (assuming they conflict with the workloads activity). Consider the cycles wasted issuing a command for data and the cycles spent transmitting the data. 
For your purposes, I think it is fine to use the theoretical bandwidth of the data bus, and not worry too much about what the actual peak bandwidth could be.

**Comment by maccoymerrell** [comment]
Also, the data bus being capable of a certaim bandwidth and being able to utilize that bandwidth are two separate things. Memory controller address mappings, scheduling policies, and bank contention are all going to play a role in how well the memory system can utilize the data bus.

**Reply by icuuuuuuer** [reply]
I fully understand your point that the maximum DDR bandwidth and the ability to utilize that bandwidth are two different things. I noticed that the develop version of Champsim has made extensive changes to the DDR-related code. Does this mean that the improved memory controller is now better and more accurate? Would you recommend using the develop version for this bandwidth-sensitive experiment? Thank you!

**Reply by maccoymerrell** [reply]
I would. I would also recommend modifying the dram address mapping in the dram controller to reflect more recent address mappings.

**Reply by icuuuuuuer** [reply]
Thank u very much.That is very helpful
